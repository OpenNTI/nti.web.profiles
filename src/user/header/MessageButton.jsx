import React from 'react';
import PropTypes from 'prop-types';

import { Flyout } from '@nti/web-commons';
import { Button } from '@nti/web-core';
import { scoped } from '@nti/lib-locale';
import { LinkTo } from '@nti/web-routing';

const t = scoped('nti-web-profiles.user.header.ManageControls', {
	message: 'Message',
	follow: 'Follow',
	unfollow: 'Unfollow',
	isOffline: '%(name)s is offline',
});

export default class ManageControls extends React.Component {
	static propTypes = {
		entity: PropTypes.object.isRequired,
		displayName: PropTypes.string.isRequired,
		presence: PropTypes.object,
	};

	render() {
		const { entity, displayName, presence } = this.props;

		if (presence?.isOnline()) {
			return (
				<LinkTo.Object context="open-chat" object={entity}>
					<Button className="message available">
						<div className="icon" />
						<span className="text">{t('message')}</span>
					</Button>
				</LinkTo.Object>
			);
		} else {
			return (
				<Flyout.Triggered
					className="message-button-flyout"
					trigger={
						<Button className="message">
							<span>
								<div className="icon" />
								<span className="text">{t('message')}</span>
							</span>
						</Button>
					}
					verticalAlign={Flyout.ALIGNMENTS.TOP}
					horizontalAlign={Flyout.ALIGNMENTS.CENTER}
					ref={this.attachFlyoutRef}
					hover
					arrow
				>
					<div className="tooltip-text">
						<span>{t('isOffline', { name: displayName })}</span>
					</div>
				</Flyout.Triggered>
			);
		}
	}
}
