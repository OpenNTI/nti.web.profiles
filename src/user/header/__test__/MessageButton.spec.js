import React from 'react';
import { act, create } from 'react-test-renderer';

import * as TestUtils from '@nti/web-client/test-utils';

import MessageButton from '../MessageButton';

const { tearDownTestClient, setupTestClient } = TestUtils;

const getMockService = () => {
	return {};
};

const onBefore = () => {
	jest.useFakeTimers();
	setupTestClient(getMockService());
};

const onAfter = () => {
	tearDownTestClient();
};

/* eslint-env jest */
describe('User profile message button test', () => {
	beforeEach(() => onBefore());
	afterEach(onAfter);

	test('Test not present', async () => {
		const user = {
			Username: 'testUser',
		};

		const presence = {
			status: null,
		};

		let cmp;
		act(() => {
			cmp = create(
				<MessageButton
					entity={user}
					displayName="TEST USER"
					presence={presence}
				/>
			);
		});

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});

	test('Test present', async () => {
		const user = {
			Username: 'testUser',
		};

		const presence = {
			status: '',
		};

		let cmp;
		act(() => {
			cmp = create(
				<MessageButton
					entity={user}
					displayName="TEST USER"
					presence={presence}
				/>
			);
		});

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
		cmp.unmount();
	});
});
