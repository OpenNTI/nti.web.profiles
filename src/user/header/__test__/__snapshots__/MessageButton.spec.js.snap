// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User profile message button test Test not present 1`] = `
<a
  className="nti-button flyout-trigger message message-button-flyout flyout-closed button primary"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  tabIndex={0}
>
  <span>
    <div
      className="icon"
    />
    <span
      className="text"
    >
      Message
    </span>
  </span>
</a>
`;

exports[`User profile message button test Test present 1`] = `
<div
  className="nti-link-to-path"
  to="#"
>
  <a
    className="nti-button message available button primary"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="icon"
    />
    <span
      className="text"
    >
      Message
    </span>
  </a>
</div>
`;
